src/coordination/coordinator_instance.cpp

```diff
@@ -1144,304 +742,241 @@ auto CoordinatorInstance::AddCoordinatorInstance(CoordinatorToCoordinatorConfig
   return AddCoordinatorInstanceStatus::SUCCESS;
 }
 
-void CoordinatorInstance::MainFailCallback(std::string_view repl_instance_name) {
-  spdlog::trace("Instance {} performing main fail callback in thread {}", repl_instance_name,
-                std::this_thread::get_id());
-
+void CoordinatorInstance::InstanceSuccessCallback(std::string_view instance_name,
+                                                  std::optional<InstanceState> instance_state) {
   if (raft_state_->IsLockOpened()) {
-    spdlog::trace("Returning from MainFailCallback since the lock is still opened.");
+    spdlog::trace("Lock is still opened, not executing instance success callback.");
     return;
   }
 
-  if (!is_leader_ready_.load(std::memory_order_acquire)) {
-    spdlog::trace("Returning from MainFailCallback since the leader is not ready.");
-    return;
-  }
-
-  auto &repl_instance = FindReplicationInstance(repl_instance_name);
-  repl_instance.OnFailPing();
-
-  // NOLINTNEXTLINE
-  if (!repl_instance.IsAlive() && raft_state_->IsCurrentMain(repl_instance_name)) {
-    spdlog::info("Cluster without main instance, trying failover.");
-    TryFailover();
-  }
-  spdlog::trace("Instance {} successfully executed on failure main callback.", repl_instance_name);
-}
-
-void CoordinatorInstance::MainSuccessCallback(std::string_view repl_instance_name) {
-  spdlog::trace("Instance {} performing main successful callback in thread", repl_instance_name,
-                std::this_thread::get_id());
-
-  if (raft_state_->IsLockOpened()) {
-    spdlog::trace("Returning from MainSuccessCallback since the lock is still opened.");
-    return;
-  }
-
-  if (!is_leader_ready_.load(std::memory_order_acquire)) {
-    spdlog::trace("Returning from MainSuccessCallback since the leader is not ready.");
-    return;
-  }
-
-  auto &repl_instance = FindReplicationInstance(repl_instance_name);
-
-  // If the instance was alive and ping successful, instance is still considered alive.
-  if (repl_instance.IsAlive()) {
-    repl_instance.OnSuccessPing();
-    spdlog::trace(
-        "Instance {} successfully executed on success main callback. Instance was alive and is still considered alive.",
-        repl_instance_name);
+  if (!is_leader_ready_) {
+    spdlog::trace("Leader is not ready, not executing instance success callback.");
     return;
   }
 
-  // NOLINTNEXTLINE
-  // The instance wasn't alive but the ping successfully passed now. If the instance is current main, send rpc message
-  // enabling writing.
-  if (raft_state_->IsCurrentMain(repl_instance.InstanceName())) {
-    spdlog::trace(
-        "Instance {} wasn't considered alive but the health ping passed now. Instance considered current main, trying "
-        "to enable writing.",
-        repl_instance_name);
-    if (!repl_instance.EnableWritingOnMain()) {
-      spdlog::error("Failed to enable writing on instance {}.", repl_instance_name);
-      return;
+  auto lock = std::unique_lock{coord_instance_lock_};
```

Comment:
Shouldn't this lock be before the ready and locked checks?

Reply:
is_leader_ready is atomic and raft_state is completely unconnected with coord_instance_lock_

Reply:
Depends on how the lock is serializing the commands.
is_leader_ready is not protected under the lock beacuse we need to change it quickly inside the reaft logs.
IsLockOpen should be protected under the lock, since otherwise you could check it, then wait for the unique lock; some other thread is holding the unique lock and changes the IsLockOpen. When you finally get the unique_lock, the IsLockOpen has changed.